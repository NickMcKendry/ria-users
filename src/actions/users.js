import database from '../firebase/firebase';

// ADD_USER

export const addUser = (user) => ({
  type: 'ADD_USER',
  user
});

export const startAddUser = (userData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      firstName = '',
      lastName = '',
      address = 0,
      createdAt = 0
    } = userData;

    const user = {
      firstName,
      lastName,
      address,
      createdAt
    };

    return database.ref(`users/${uid}/users`).push(user).then((ref) => {
      dispatch(addUser({
        id: ref.key,
        ...user
      }));
    });
  };
};

// REMOVE_USER

export const removeUser = ({ id } = {}) => ({
  type: 'REMOVE_USER',
  id
});

export const startRemoveUser = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users/${id}`).remove().then(() => {
      dispatch(removeUser({
        id
      }))
    })
  }
};

// EDIT_USER

export const editUser = (id, updates) => ({
  type: 'EDIT_USER',
  id,
  updates
});

export const startEditUser = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users/${id}`).update(updates).then(() => {
      dispatch(editUser(id, updates))
    });
  };
};

// SET_USERS

export const setUsers = (users) => ({
  type: 'SET_USERS',
  users
});

export const startSetUsers = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users`).once('value').then((snapshot) => {
      const users = [];

      snapshot.forEach((childSnapshot) => {
        users.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setUsers(users));
    });
  };
};
