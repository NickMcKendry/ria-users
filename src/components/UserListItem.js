import React from 'react';
import { Link } from 'react-router-dom';

const UserListItem = ({ id, firstName, lastName, address, createdAt }) => (
    <Link to={`/edit/${id}`} className="list-item">
      <div>
        <h3 className="list-item__title">{firstName} {lastName}</h3>
      </div>
      <h3 className="list-item__data">{address}</h3>
    </Link>
);

export default UserListItem;
