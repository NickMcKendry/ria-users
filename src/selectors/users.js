import moment from 'moment';

export default (users, { text, sortBy, birthDate, endDate }) => {

  return users.filter((user) => {
    const createdAtMoment = moment(user.createdAt)

    const birthDateMatch = birthDate ? birthDate.isSameOrBefore(createdAtMoment, 'day') : true;

    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

    const name = user.firstName + user.lastName

    const textMatch = name.toLowerCase().includes(text.toLowerCase());

    return birthDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if(sortBy === 'first-name') {
      return a.amount < b.amount ? 1 : -1;
    } else if(sortBy === 'last-name') {
      return a.amount < b.amount ? 1 : -1;
    } else if(sortBy === 'address') {
      return a.amount < b.amount ? 1 : -1;
    }
  })
}
