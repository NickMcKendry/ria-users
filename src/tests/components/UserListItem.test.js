import React from 'react';
import UserListItem from '../../components/UserListItem';
import { shallow } from 'enzyme';
import users from '../fixtures/users';

test('Should render user list item', () => {
  const wrapper = shallow(<UserListItem {...users[0]} />);
  expect(wrapper).toMatchSnapshot();
});
