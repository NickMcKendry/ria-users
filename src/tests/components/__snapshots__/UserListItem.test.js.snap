// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render user list item 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Nick
       
      McKendry
    </h3>
  </div>
  <h3
    className="list-item__data"
  >
    9803 w. 83rd ave.
  </h3>
</Link>
`;
